// @cardano-sdk/core/dist/cjs/Serialization/Scripts/NativeScript/TimelockStart.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TimelockStart_slot, _TimelockStart_originalBytes;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelockStart = void 0;
const Cardano = __importStar(require("../../../Cardano"));
const CBOR_1 = require("../../CBOR");
const util_1 = require("@cardano-sdk/util");
const EMBEDDED_GROUP_SIZE = 2;
class TimelockStart {
    constructor(slot) {
        _TimelockStart_slot.set(this, void 0);
        _TimelockStart_originalBytes.set(this, undefined);
        __classPrivateFieldSet(this, _TimelockStart_slot, slot, "f");
    }
    toCbor() {
        if (__classPrivateFieldGet(this, _TimelockStart_originalBytes, "f"))
            return __classPrivateFieldGet(this, _TimelockStart_originalBytes, "f");
        const writer = new CBOR_1.CborWriter();
        writer.writeStartArray(EMBEDDED_GROUP_SIZE);
        writer.writeInt(Cardano.NativeScriptKind.RequireTimeAfter);
        writer.writeInt(__classPrivateFieldGet(this, _TimelockStart_slot, "f"));
        return writer.encodeAsHex();
    }
    static fromCbor(cbor) {
        const reader = new CBOR_1.CborReader(cbor);
        const length = reader.readStartArray();
        if (length !== EMBEDDED_GROUP_SIZE)
            throw new util_1.InvalidArgumentError('cbor', `Expected an array of two elements, but got an array of ${length} elements`);
        const kind = Number(reader.readInt());
        if (kind !== Cardano.NativeScriptKind.RequireTimeAfter)
            throw new util_1.InvalidArgumentError('cbor', `Expected kind ${Cardano.NativeScriptKind.RequireTimeAfter}, but got kind ${kind}`);
        const slot = Cardano.Slot(Number(reader.readInt()));
        const script = new TimelockStart(slot);
        __classPrivateFieldSet(script, _TimelockStart_originalBytes, cbor, "f");
        return script;
    }
    toCore() {
        return {
            __type: Cardano.ScriptType.Native,
            kind: Cardano.NativeScriptKind.RequireTimeAfter,
            slot: __classPrivateFieldGet(this, _TimelockStart_slot, "f")
        };
    }
    static fromCore(script) {
        return new TimelockStart(script.slot);
    }
    slot() {
        return __classPrivateFieldGet(this, _TimelockStart_slot, "f");
    }
    setSlot(slot) {
        __classPrivateFieldSet(this, _TimelockStart_slot, slot, "f");
        __classPrivateFieldSet(this, _TimelockStart_originalBytes, undefined, "f");
    }
}
exports.TimelockStart = TimelockStart;
_TimelockStart_slot = new WeakMap(), _TimelockStart_originalBytes = new WeakMap();
