// @cardano-sdk/util-dev/dist/cjs/assetId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.All = exports.E = exports.D = exports.C = exports.B = exports.A = exports.Unit = exports.PXL = exports.TSLA = void 0;
const core_1 = require("@cardano-sdk/core");
exports.TSLA = core_1.Cardano.AssetId('659f2917fb63f12b33667463ee575eeac1845bbc736b9c0bbc40ba8254534c41');
exports.PXL = core_1.Cardano.AssetId('1ec85dcee27f2d90ec1f9a1e4ce74a667dc9be8b184463223f9c960150584c');
exports.Unit = core_1.Cardano.AssetId('a5425bd7bc4182325188af2340415827a73f845846c165d9e14c5aed556e6974');
exports.A = core_1.Cardano.AssetId('659f2917fb63f12b33667463ee575eeac1845bbc736b9c0bbc40ba8241');
exports.B = core_1.Cardano.AssetId('659f2917fb63f12b33667463ee575eeac1845bbc736b9c0bbc40ba8242');
exports.C = core_1.Cardano.AssetId('659f2917fb63f12b33667463ee575eeac1845bbc736b9c0bbc40ba8243');
exports.D = core_1.Cardano.AssetId('659f2917fb63f12b33667463ee575eeac1845bbc736b9c0bbc40ba8244');
exports.E = core_1.Cardano.AssetId('659f2917fb63f12b33667463ee575eeac1845bbc736b9c0bbc40ba8245');
exports.All = [exports.TSLA, exports.PXL, exports.Unit];
