// @cardano-sdk/util-dev/dist/cjs/Cip19TestVectors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.byronTestnetDaedalus = exports.byronMainnetYoroi = exports.testnetRewardScript = exports.testnetRewardKey = exports.testnetEnterpriseScript = exports.testnetEnterpriseKey = exports.testnetPointerScript = exports.testnetPointerKey = exports.testnetBasePaymentScriptStakeScript = exports.testnetBasePaymentKeyStakeScript = exports.testnetBasePaymentScriptStakeKey = exports.testnetBasePaymentKeyStakeKey = exports.rewardScript = exports.rewardKey = exports.enterpriseScript = exports.enterpriseKey = exports.pointerScript = exports.pointerKey = exports.basePaymentScriptStakeScript = exports.basePaymentKeyStakeScript = exports.basePaymentScriptStakeKey = exports.basePaymentKeyStakeKey = exports.POINTER = exports.SCRIPT_CREDENTIAL = exports.KEY_STAKE_CREDENTIAL = exports.KEY_PAYMENT_CREDENTIAL = exports.SCRIPT_HASH = exports.STAKE_KEY_HASH = exports.PAYMENT_KEY_HASH = void 0;
const core_1 = require("@cardano-sdk/core");
exports.PAYMENT_KEY_HASH = '9493315cd92eb5d8c4304e67b7e16ae36d61d34502694657811a2c8e';
exports.STAKE_KEY_HASH = '337b62cfff6403a06a3acbc34f8c46003c69fe79a3628cefa9c47251';
exports.SCRIPT_HASH = 'c37b1b5dc0669f1d3c61a6fddb2e8fde96be87b881c60bce8e8d542f';
exports.KEY_PAYMENT_CREDENTIAL = {
    hash: exports.PAYMENT_KEY_HASH,
    type: core_1.Cardano.CredentialType.KeyHash
};
exports.KEY_STAKE_CREDENTIAL = {
    hash: exports.STAKE_KEY_HASH,
    type: core_1.Cardano.CredentialType.KeyHash
};
exports.SCRIPT_CREDENTIAL = {
    hash: exports.SCRIPT_HASH,
    type: core_1.Cardano.CredentialType.ScriptHash
};
exports.POINTER = {
    certIndex: core_1.Cardano.CertIndex(3),
    slot: core_1.Cardano.Slot(2498243),
    txIndex: core_1.Cardano.TxIndex(27)
};
exports.basePaymentKeyStakeKey = 'addr1qx2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer3n0d3vllmyqwsx5wktcd8cc3sq835lu7drv2xwl2wywfgse35a3x';
exports.basePaymentScriptStakeKey = 'addr1z8phkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gten0d3vllmyqwsx5wktcd8cc3sq835lu7drv2xwl2wywfgs9yc0hh';
exports.basePaymentKeyStakeScript = 'addr1yx2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzerkr0vd4msrxnuwnccdxlhdjar77j6lg0wypcc9uar5d2shs2z78ve';
exports.basePaymentScriptStakeScript = 'addr1x8phkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gt7r0vd4msrxnuwnccdxlhdjar77j6lg0wypcc9uar5d2shskhj42g';
exports.pointerKey = 'addr1gx2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer5pnz75xxcrzqf96k';
exports.pointerScript = 'addr128phkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gtupnz75xxcrtw79hu';
exports.enterpriseKey = 'addr1vx2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzers66hrl8';
exports.enterpriseScript = 'addr1w8phkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gtcyjy7wx';
exports.rewardKey = 'stake1uyehkck0lajq8gr28t9uxnuvgcqrc6070x3k9r8048z8y5gh6ffgw';
exports.rewardScript = 'stake178phkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gtcccycj5';
exports.testnetBasePaymentKeyStakeKey = 'addr_test1qz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer3n0d3vllmyqwsx5wktcd8cc3sq835lu7drv2xwl2wywfgs68faae';
exports.testnetBasePaymentScriptStakeKey = 'addr_test1zrphkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gten0d3vllmyqwsx5wktcd8cc3sq835lu7drv2xwl2wywfgsxj90mg';
exports.testnetBasePaymentKeyStakeScript = 'addr_test1yz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzerkr0vd4msrxnuwnccdxlhdjar77j6lg0wypcc9uar5d2shsf5r8qx';
exports.testnetBasePaymentScriptStakeScript = 'addr_test1xrphkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gt7r0vd4msrxnuwnccdxlhdjar77j6lg0wypcc9uar5d2shs4p04xh';
exports.testnetPointerKey = 'addr_test1gz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzer5pnz75xxcrdw5vky';
exports.testnetPointerScript = 'addr_test12rphkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gtupnz75xxcryqrvmw';
exports.testnetEnterpriseKey = 'addr_test1vz2fxv2umyhttkxyxp8x0dlpdt3k6cwng5pxj3jhsydzerspjrlsz';
exports.testnetEnterpriseScript = 'addr_test1wrphkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gtcl6szpr';
exports.testnetRewardKey = 'stake_test1uqehkck0lajq8gr28t9uxnuvgcqrc6070x3k9r8048z8y5gssrtvn';
exports.testnetRewardScript = 'stake_test17rphkx6acpnf78fuvxn0mkew3l0fd058hzquvz7w36x4gtcljw6kf';
exports.byronMainnetYoroi = 'Ae2tdPwUPEZFRbyhz3cpfC2CumGzNkFBN2L42rcUc2yjQpEkxDbkPodpMAi';
exports.byronTestnetDaedalus = '37btjrVyb4KEB2STADSsj3MYSAdj52X5FrFWpw2r7Wmj2GDzXjFRsHWuZqrw7zSkwopv8Ci3VWeg6bisU9dgJxW5hb2MZYeduNKbQJrqz3zVBsu9nT';
