// @cardano-sdk/key-management/dist/cjs/KeyAgentBase.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _KeyAgentBase_serializableData, _KeyAgentBase_bip32Ed25519, _KeyAgentBase_account;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyAgentBase = void 0;
const types_1 = require("./types");
const Bip32Account_1 = require("./Bip32Account");
class KeyAgentBase {
    constructor(serializableData, { bip32Ed25519 }) {
        _KeyAgentBase_serializableData.set(this, void 0);
        _KeyAgentBase_bip32Ed25519.set(this, void 0);
        _KeyAgentBase_account.set(this, void 0);
        __classPrivateFieldSet(this, _KeyAgentBase_serializableData, serializableData, "f");
        __classPrivateFieldSet(this, _KeyAgentBase_bip32Ed25519, bip32Ed25519, "f");
        __classPrivateFieldSet(this, _KeyAgentBase_account, new Bip32Account_1.Bip32Account(serializableData), "f");
    }
    get serializableData() {
        return __classPrivateFieldGet(this, _KeyAgentBase_serializableData, "f");
    }
    get extendedAccountPublicKey() {
        return this.serializableData.extendedAccountPublicKey;
    }
    get chainId() {
        return this.serializableData.chainId;
    }
    get accountIndex() {
        return this.serializableData.accountIndex;
    }
    get bip32Ed25519() {
        return __classPrivateFieldGet(this, _KeyAgentBase_bip32Ed25519, "f");
    }
    get purpose() {
        return this.serializableData.purpose || types_1.KeyPurpose.STANDARD;
    }
    async deriveAddress(paymentKeyDerivationPath, stakeKeyDerivationIndex) {
        return __classPrivateFieldGet(this, _KeyAgentBase_account, "f").deriveAddress(paymentKeyDerivationPath, stakeKeyDerivationIndex);
    }
    async derivePublicKey(derivationPath) {
        return (await __classPrivateFieldGet(this, _KeyAgentBase_account, "f").derivePublicKey(derivationPath)).hex();
    }
}
exports.KeyAgentBase = KeyAgentBase;
_KeyAgentBase_serializableData = new WeakMap(), _KeyAgentBase_bip32Ed25519 = new WeakMap(), _KeyAgentBase_account = new WeakMap();
