// @cardano-sdk/key-management/dist/cjs/util/bip39.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateMnemonic = exports.mnemonicWordsToEntropy = exports.entropyToMnemonicWords = exports.joinMnemonicWords = exports.generateMnemonicWords = exports.mnemonicToWords = void 0;
const bip39 = __importStar(require("bip39"));
const mnemonicToWords = (mnemonic) => mnemonic.split(' ');
exports.mnemonicToWords = mnemonicToWords;
const generateMnemonicWords = (strength = 256) => (0, exports.mnemonicToWords)(bip39.generateMnemonic(strength));
exports.generateMnemonicWords = generateMnemonicWords;
const joinMnemonicWords = (mnenomic) => mnenomic.join(' ');
exports.joinMnemonicWords = joinMnemonicWords;
const entropyToMnemonicWords = (entropy) => (0, exports.mnemonicToWords)(bip39.entropyToMnemonic(entropy));
exports.entropyToMnemonicWords = entropyToMnemonicWords;
const mnemonicWordsToEntropy = (mnenonic) => bip39.mnemonicToEntropy((0, exports.joinMnemonicWords)(mnenonic));
exports.mnemonicWordsToEntropy = mnemonicWordsToEntropy;
exports.validateMnemonic = bip39.validateMnemonic;
