// @cardano-sdk/util-dev/dist/cjs/mockProviders/mockNetworkInfoProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockNetworkInfoProvider = exports.networkInfo = void 0;
const mockData_1 = require("./mockData");
const eraSummaries_1 = require("../eraSummaries");
exports.networkInfo = {
    lovelaceSupply: {
        circulating: 42064399450423723n,
        total: 40267211394073980n
    },
    network: {
        eraSummaries: eraSummaries_1.testnetEraSummaries,
        magic: 1097911063
    },
    stake: {
        active: 1060378314781343n,
        live: 15001884895856815n
    }
};
const mockNetworkInfoProvider = () => ({
    eraSummaries: jest.fn().mockResolvedValue(exports.networkInfo.network.eraSummaries),
    genesisParameters: jest.fn().mockResolvedValue(mockData_1.genesisParameters),
    healthCheck: jest.fn().mockResolvedValue({ ok: true }),
    ledgerTip: jest.fn().mockResolvedValue(mockData_1.ledgerTip),
    lovelaceSupply: jest.fn().mockResolvedValue(exports.networkInfo.lovelaceSupply),
    protocolParameters: jest.fn().mockResolvedValue(mockData_1.protocolParameters),
    stake: jest.fn().mockResolvedValue(exports.networkInfo.stake)
});
exports.mockNetworkInfoProvider = mockNetworkInfoProvider;
