// @cardano-sdk/core/dist/cjs/Serialization/Update/Costmdls/CostModel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _CostModel_instances, _CostModel_language, _CostModel_costs, _CostModel_isOperationValid;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CostModel = void 0;
const Cardano = __importStar(require("../../../Cardano"));
const util_1 = require("@cardano-sdk/util");
const PLUTUS_V1_COST_MODEL_OP_COUNT = 166;
const PLUTUS_V2_COST_MODEL_OP_COUNT = 175;
const PLUTUS_V3_COST_MODEL_OP_COUNT = 179;
class CostModel {
    constructor(language, costs) {
        _CostModel_instances.add(this);
        _CostModel_language.set(this, void 0);
        _CostModel_costs.set(this, void 0);
        __classPrivateFieldSet(this, _CostModel_language, language, "f");
        __classPrivateFieldSet(this, _CostModel_costs, costs, "f");
        switch (__classPrivateFieldGet(this, _CostModel_language, "f")) {
            case Cardano.PlutusLanguageVersion.V1:
                if (costs.length !== PLUTUS_V1_COST_MODEL_OP_COUNT)
                    throw new util_1.InvalidArgumentError('costs', `Cost model for PlutusV2 language should have ${PLUTUS_V2_COST_MODEL_OP_COUNT} operations, but got ${costs.length}.`);
                break;
            case Cardano.PlutusLanguageVersion.V2:
                if (costs.length !== PLUTUS_V2_COST_MODEL_OP_COUNT)
                    throw new util_1.InvalidArgumentError('costs', `Cost model for PlutusV2 language should have ${PLUTUS_V2_COST_MODEL_OP_COUNT} operations, but got ${costs.length}.`);
                break;
            case Cardano.PlutusLanguageVersion.V3:
                if (costs.length !== PLUTUS_V3_COST_MODEL_OP_COUNT)
                    throw new util_1.InvalidArgumentError('costs', `Cost model for PlutusV3 language should have ${PLUTUS_V3_COST_MODEL_OP_COUNT} operations, but got ${costs.length}.`);
                break;
            default:
                throw new util_1.InvalidStateError('Invalid plutus language version.');
        }
    }
    static newPlutusV1(costs) {
        return new CostModel(Cardano.PlutusLanguageVersion.V1, costs);
    }
    static newPlutusV2(costs) {
        return new CostModel(Cardano.PlutusLanguageVersion.V2, costs);
    }
    static newPlutusV3(costs) {
        return new CostModel(Cardano.PlutusLanguageVersion.V3, costs);
    }
    set(operation, cost) {
        if (!__classPrivateFieldGet(this, _CostModel_instances, "m", _CostModel_isOperationValid).call(this, operation))
            throw new util_1.InvalidArgumentError('operation', `The given operation ${operation} is invalid for the current Language version ${__classPrivateFieldGet(this, _CostModel_language, "f")}.`);
        __classPrivateFieldGet(this, _CostModel_costs, "f")[operation] = cost;
    }
    get(operation) {
        if (!__classPrivateFieldGet(this, _CostModel_instances, "m", _CostModel_isOperationValid).call(this, operation))
            throw new util_1.InvalidArgumentError('operation', `The given operation ${operation} is invalid for the current Language version ${__classPrivateFieldGet(this, _CostModel_language, "f")}.`);
        return __classPrivateFieldGet(this, _CostModel_costs, "f")[operation];
    }
    costs() {
        return __classPrivateFieldGet(this, _CostModel_costs, "f");
    }
    language() {
        return __classPrivateFieldGet(this, _CostModel_language, "f");
    }
}
exports.CostModel = CostModel;
_CostModel_language = new WeakMap(), _CostModel_costs = new WeakMap(), _CostModel_instances = new WeakSet(), _CostModel_isOperationValid = function _CostModel_isOperationValid(operation) {
    let isValid = false;
    switch (__classPrivateFieldGet(this, _CostModel_language, "f")) {
        case Cardano.PlutusLanguageVersion.V1:
            isValid = operation >= 0 && operation < PLUTUS_V1_COST_MODEL_OP_COUNT;
            break;
        case Cardano.PlutusLanguageVersion.V2:
            isValid = operation >= 0 && operation < PLUTUS_V2_COST_MODEL_OP_COUNT;
            break;
        case Cardano.PlutusLanguageVersion.V3:
            isValid = operation >= 0 && operation < PLUTUS_V3_COST_MODEL_OP_COUNT;
            break;
        default:
            throw new util_1.InvalidStateError('Invalid plutus language version.');
    }
    return isValid;
};
